src/tasks/cardid-task/cardid-task.c

```diff
@@ -0,0 +1,138 @@
+#include "cardid-task.h"
+
+/*===========================================================================*/
+/* Internal data structures and constants                                    */
+/*===========================================================================*/
+
+#define THREAD_WORKING_AREA_SIZE 1024
+
+/*===========================================================================*/
+/* Internal variables                                                        */
+/*===========================================================================*/
+
+static uint8_t                      task_id;
+
+static volatile bool                poll = false;
+static mutex_t                      pollMutex;
+
+static dl_task_cardid_callbacks     *callbacks;
+
+static THD_WORKING_AREA(cardIDTaskThreadWA, THREAD_WORKING_AREA_SIZE);
+static thread_t                     *task_thread;
+
+
+/*===========================================================================*/
+/* Task-specific API                                                         */
+/*===========================================================================*/
+
+void dlTaskCardIDStartPolling(void) {
+    chMtxLock(&pollMutex);
+    poll = true;
+    chMtxUnlock(&pollMutex);
+}
+
+void dlTaskCardIDStopPolling(void) {
+    chMtxLock(&pollMutex);
+    poll = false;
+    chMtxUnlock(&pollMutex);
+}
+
+
+/*===========================================================================*/
+/* Task thread and internal functions                                        */
+/*===========================================================================*/
+
+static THD_FUNCTION(cardIDTask, arg) {
+    (void) arg;
+    bool readerActive = false;
+
+    while (!chThdShouldTerminateX()) {
+        // Mutex lock here is not necessary. If a race condition occurs the worst case is that the
+        // RFID Reader will perform one more poll cycle.
+        // If the poll changes to false during polling, it is checked again (with mutex acquired)
+        // just before invoking a callback.
+        bool pollThisCycle = poll;
+
+        // Activate / deactivate reader
+        if (pollThisCycle != readerActive) {
+            pcdresult_t result;
+            if (pollThisCycle) {
+                result = pcdActivateRFAB(PCD);
+            } else {
+                result = pcdDeactivateRFAB(PCD);
+            }
+
+            if (result != PCD_OK) {
+                chMtxLock(&pollMutex);
+                callbacks->reader_error();
+                poll = false;
+                readerActive = false;
+                // The module is in unknown state, best to try and reset it so that we know that
+                // RF field is inactive.
+                resetRFIDModule();
+                chMtxUnlock(&pollMutex);
+            } else {
+                readerActive = pollThisCycle;
+            }
+        }
+
+        if (pollThisCycle) {
+            unsigned int result_p = 0;
+            Picc cards[10];
```

Comment:
I see there's a maximum of 10 cards per poll. This number is used in multiple places, so defining the constant would be appropriate.

Reply:
Fixed